File: BOJ/basic-algo-lecture/0x10 다이나믹 프로그래밍/1915.cpp
#include <bits/stdc++.h>
using namespace std;

int board[1005][1005];
int d[1005][1005];

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    int n, m;
    cin >> n >> m;

    for (int i = 0; i < n; i++) {
        string s;
        cin >> s;
        for (int j = 0; j < m; j++) board[i][j] = s[j] - '0';
    }

    int ans = 0;
    for (int i = 0; i < n; i++)
        for (int j = 0; j < m; j++) {
            if (i == 0 || j == 0) d[i][j] = board[i][j];
            else if (board[i][j]) d[i][j] = min(d[i - 1][j - 1], min(d[i - 1][j], d[i][j - 1])) + 1;
            ans = max(ans, d[i][j]);
        }
    cout << ans * ans;
}

File: 바킹독의 실전 알고리즘/0x10강 다이나믹 프로그래밍/1915.py
import sys
input = lambda: sys.stdin.readline().rstrip()
mis = lambda: map(int, input().split())
ii = lambda: int(input())

n, m = mis()
board = [list(map(int, input())) for _ in range(n)]
d = [[0] * m for _ in range(n)]
ans = 0
for i in range(n):
    for j in range(m):
        if i == 0 or j == 0:
            d[i][j] = board[i][j]
        elif board[i][j]:
            d[i][j] = min(d[i - 1][j - 1], d[i - 1][j], d[i][j - 1]) + 1
        ans = max(ans, d[i][j])
print(ans * ans)